File: app/src/org/koreader/launcher/device/EPDFactory.java
/* generic EPD Controller for Android devices,
 * based on https://github.com/unwmun/refreshU */

package org.koreader.launcher.device;

import java.util.Locale;

import org.koreader.launcher.Logger;
import org.koreader.launcher.device.rockchip.RK3026EPDController;
import org.koreader.launcher.device.rockchip.RK3066EPDController;
import org.koreader.launcher.device.rockchip.RK3368EPDController;
import org.koreader.launcher.device.freescale.NTXNewEPDController;


public class EPDFactory {

    public static EPDController getEPDController() {
        EPDController epdController = null;
        String controllerName = null;

        switch (DeviceInfo.EINK) {

            /* Supported rk3026 devices */
            case BOYUE_T61:
            case BOYUE_T80S:
            case ONYX_C67:
            case ENERGY:
            case INKBOOK:
                controllerName = "Rockchip RK3026";
                epdController = new RK3026EPDController();
                break;

            /* supported rk3066 devices */
            case BOYUE_T62:
                controllerName = "Rockchip RK3066";
                epdController = new RK3066EPDController();
                break;

            /* supported rk3368 devices */
            case BOYUE_T80D:
            case BOYUE_T78D:
            case BOYUE_T103D:
                controllerName = "Rockchip RK3368";
                epdController = new RK3368EPDController();
                break;

            /* devices using imx/ntx platform */
            case CREMA:
            case TOLINO:
            case NOOK_V520:
                controllerName = "Freescale NTX";
                epdController = new NTXNewEPDController();
                break;

            /* unsupported devices */
            case UNKNOWN:
                epdController = new FakeEPDController();
                break;

            default : break;
        }

        if (controllerName != null) {
            Logger.i(String.format(Locale.US,
                "[EPDFactory]: Using %s EPD Controller", controllerName));
        }

        return epdController;
    }

    private static class FakeEPDController implements EPDController {
        @Override
        public void setEpdMode(android.view.View targetView,
                               int mode, long delay,
                               int x, int y, int width, int height, String epdMode)
        {
            Logger.w("[EPDController]: Unknown device, ignoring epd update");
        }
    }
}

File: app/src/org/koreader/launcher/EPDFactory.kt
/* generic EPD Controller for Android devices,
 * based on https://github.com/unwmun/refreshU */

package org.koreader.launcher

import java.util.Locale

import org.koreader.launcher.device.rockchip.RK3026EPDController
import org.koreader.launcher.device.rockchip.RK3066EPDController
import org.koreader.launcher.device.rockchip.RK3368EPDController
import org.koreader.launcher.device.freescale.NTXNewEPDController

internal object EPDFactory {
    val epdController: EPDController
        get() {
            return when (DeviceInfo.EINK) {
                DeviceInfo.EinkDevice.BOYUE_T61,
                DeviceInfo.EinkDevice.BOYUE_T62,
                DeviceInfo.EinkDevice.BOYUE_T80S,
                DeviceInfo.EinkDevice.ONYX_C67,
                DeviceInfo.EinkDevice.ENERGY,
                DeviceInfo.EinkDevice.INKBOOK -> {
                    logController("Rockchip RK3026")
                    RK3026EPDController()
                }
                DeviceInfo.EinkDevice.BOYUE_T80D,
                DeviceInfo.EinkDevice.BOYUE_T78D,
                DeviceInfo.EinkDevice.BOYUE_T103D -> {
                    logController("Rockchip RK3368")
                    RK3368EPDController()
                }
                DeviceInfo.EinkDevice.CREMA,
                DeviceInfo.EinkDevice.TOLINO,
                DeviceInfo.EinkDevice.NOOK_V520 -> {
                    logController("Freescale NTX")
                    NTXNewEPDController()
                }
                else -> {
                    FakeEPDController()
                }
            }
        }

    private fun logController(name: String?) {
        Logger.i(String.format(Locale.US,
            "[EPDFactory]: Using %s EPD Controller", name))
    }
    private class FakeEPDController : EPDController {
        override fun setEpdMode(targetView: android.view.View,
                                mode: Int, delay: Long,
                                x: Int, y: Int, width: Int, height: Int, epdMode: String?) {
            Logger.w("[EPDController]: Unknown device, ignoring epd update")
        }
    }
}
